### PR TITLE
moved to tox, added poetry and reformatted some files.

### LINKED ISSUE #49: As a developer I would like fasjson to be of a standard format so that it is easier to develop, manage and deploy
## Acceptance Criteria

* The project should follow a standard format
* The project should use a dependency manager
* The project should use `tox` to make managing and running common commands easier
* The dependencies should be clearly defined along with any other tools in a `pyproject.toml` file

## Definition of Done
All of the above are complete

### PR DESCRIPTION
**Pull request by StephenCoady**
Signed-off-by: Stephen Coady <scoady@redhat.com>

Fixes #49 

## Proposed Changes

* 

## Verification Steps

*

### CODE REVIEW

File: .vscode/settings.json [hunk 1]
```diff
@@ -0,0 +1,3 @@
+{
```
Comment (abompard) [inline .vscode/settings.json:1]: This should probably not be committed

File: tox.ini [hunk 2]
```diff
@@ -0,0 +1,66 @@
+[tox]
+envlist = lint,format,licenses,bandit,covclean,{py36,py37}-{unittest},covreport
+isolated_build = true
+
+[testenv]
+setenv =
+    FASJSON_CONFIG_PATH={toxinidir}/fasjson/tests/fixtures/fasjson.cfg
+sitepackages = true
+commands =
+    /srv/venv/bin/poetry install -q
```
Comment (StephenCoady) [inline tox.ini:10]: the downside of this is that tox will only be able to run within the VM (or a machine set up this way). I've aliased poetry but tox ignores that.

File: ansible/roles/fasjson/files/fasjson.service [hunk 3]
```diff
@@ -0,0 +1,18 @@
+.include /lib/systemd/system/fasjson.service
```
Comment (abompard) [inline ansible/roles/fasjson/files/fasjson.service:1]: Do we use that file? I thought we were running the app through apache.

File: ansible/roles/fasjson/files/fasjson.wsgi [hunk 4]
```diff
@@ -1 +1,4 @@
+import sys
+sys.path.insert(0, '/vagrant/fasjson')
```
Comment (abompard) [inline ansible/roles/fasjson/files/fasjson.wsgi:2]: This should not be necessary if the proper `WSGIPythonHome` config is set in Apache's conf file

File: fasjson.egg-info/PKG-INFO [hunk 5]
```diff
@@ -0,0 +1,104 @@
+Metadata-Version: 1.2
```
Comment (abompard) [inline fasjson.egg-info/PKG-INFO:1]: The whole `.egg-info` directory should not be committed, it should be in `.gitignore` instead.

File: fasjson/lib/ldaputils.py [hunk 6]
```diff
@@ -1,7 +1,7 @@
 import datetime
 
 import ldap
-import ldap.filter as ldap_filter #type: ignore
+import ldap.filter as ldap_filter  # type: ignore
```
Comment (abompard) [inline fasjson/lib/ldaputils.py:4]: I usually try to make a separate commit for the `black` run, so that I don't need to look for logic change when reviewing that patch.

File: fasjson/tests/unit/web_resource_v1_root.py [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+from fasjson.web import app
+
+
+def test_root_success():
+    pytest.skip('TODO: implementation missing')
+
+
+def test_root_error():
+    pytest.skip('TODO: implementation missing')
+<<<<<<< HEAD
```
Comment (abompard) [inline fasjson/tests/unit/web_resource_v1_root.py:13]: Looks like a merge artefact.

File: fasjson/web/apis/v1/resources/root.py [hunk 8]
```diff
@@ -1,5 +1,38 @@
...
+
 
 def root():
-    raise errors.WebApiError('method not implemented', 405, data={'reason': 'method not implemented'})
+    raise errors.WebApiError(
+        'method not implemented',
+        405,
+        data={'reason': 'method not implemented'},
+    )
+<<<<<<< HEAD
```
Comment (abompard) [inline fasjson/web/apis/v1/resources/root.py:12]: Looks like a merge artefact.

File: fasjson/web/apis/v1/resources/root.py [hunk 9]
```diff
@@ -1,5 +1,38 @@
...
+    Return 200 if the app can query the db, to signify being ready.
+
+    Args:
+        request (pyramid.request.Request): The current request.
+    Returns:
+        dict: A dictionary with list of services checked.
+    """
+    return 'ok'
+=======
+>>>>>>> 66fda81... check in
```
Comment (abompard) [inline fasjson/web/apis/v1/resources/root.py:38]: Same here

File: fasjson/web/app.py [hunk 10]
```diff
@@ -13,42 +13,46 @@
 app = Flask(__name__)
 app.response_class = ApiResponse
 
-#extensions
+<<<<<<< HEAD
```
Comment (abompard) [inline fasjson/web/app.py:16]: Merge artefact again.

File: fasjson/web/defaults.cfg [hunk 11]
```diff
@@ -0,0 +1,3 @@
+KRB5CCNAME="/tmp/krb5cc-httpd"
+GSS_USE_PROXY="yes"
+GSS_NAME="admin@EXAMPLE.TEST"
```
Comment (abompard) [inline fasjson/web/defaults.cfg:3]: I would prefer we put test-related values in the test-specific config file.

File: fasjson/web/errors.py [hunk 12]
```diff
@@ -5,53 +5,49 @@
 
 
 class WebApiError(HTTPException):
-	"""
+    """
 	HTTP exception to be used across the web api.
```
Comment (abompard) [inline fasjson/web/errors.py:9]: This seems wrongly indented.

File: fasjson/web/errors.py [hunk 13]
```diff
@@ -5,53 +5,49 @@
...
-		"""
+    def __init__(self, message, code, data=None):
+        super(WebApiError, self).__init__(message)
+        self.code = code
+        self.message = message
+        self.extra = data
+
+    def as_dict(self):
+        """
 		Returns exception data in dict format.
```
Comment (abompard) [inline fasjson/web/errors.py:22]: Wrongly indented here too. More of that down this file.

File: pyproject.toml [hunk 14]
```diff
@@ -1,2 +1,143 @@
...
+pluggy = "^0.13.1"
+pyasn1 = "^0.4.8"
+pyasn1-modules = "^0.2.8"
+pyparsing = "^2.4.6"
+pytz = "^2019.3"
+six = "^1.14.0"
+termcolor = "^1.1.0"
+typed-ast = "^1.4.1"
+wcwidth = "^0.1.8"
+zipp = "^3.0.0"
```
Comment (abompard) [inline pyproject.toml:47]: Actually only the content of the `install_requires` in `setup.py` should go here, not all the sub-dependencies.

File: pyproject.toml [hunk 15]
```diff
@@ -1,2 +1,143 @@
...
+mypy-extensions = "^0.4.3"
+pytest-vcr = "1.0"
+vcrpy = "^4.0.2"
+liccheck = "^0.4.2"
+bandit = "^1.6"
+black = {version = "^18.3-alpha.0", allow-prereleases = true}
+
 [tool.black]
 line-length = 78
+skip-string-normalization = true
```
Comment (abompard) [inline pyproject.toml:66]: We can drop that config file since we're starting pretty much from scratch.

File: tox.ini [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+[testenv:lint]
+commands =
+    /srv/venv/bin/poetry install -q
+    /srv/venv/bin/poetry run flake8 {posargs}
+
+[testenv:format]
+commands =
+    /srv/venv/bin/poetry install -q
+    /srv/venv/bin/poetry run black --check {posargs:.}
+whitelist_externals =
```
Comment (abompard) [inline tox.ini:50]: This does not seem necessary

File: tox.ini (hunk 2 above)
Comment (abompard) [inline tox.ini:10]: Actually it's just `poetry` in this file since tox creates its own virtualenvs. Otherwise we can't run tox outside of vagrant.

Comment (StephenCoady) [reply tox.ini:10]: It does, but I mentioned above about poetry not being installed locally now. I've added it to the dnf install now so that tox can use it instead.

**Comment by StephenCoady** [comment]
@abompard this ended up being messy to re-organise because I made the linter changes with moving files. sorry about that. I'll try to not do that in future. I separated things out as best as I could - hopefully that doesn't make the re-review too messy.

File: tox.ini [hunk 17]
```diff
@@ -0,0 +1,72 @@
+[tox]
+envlist = lint,format,licenses,bandit,covclean,{py36,py37}-{unittest},covreport
+isolated_build = true
+
+[testenv]
+setenv =
+    FASJSON_CONFIG_PATH={toxinidir}/fasjson/tests/fixtures/fasjson.cfg
```
Comment (abompard) [inline tox.ini:7]: I dont think we use such an env variable

File: tox.ini [hunk 18]
```diff
@@ -0,0 +1,72 @@
...
+sitepackages = true
+commands =
+    poetry install -q
+    unittest: poetry run pytest -vv --cov --cov-append --cov-report= fasjson/tests/unit {posargs}
+depends =
+    {py36,py37}: covclean
+    covreport: py36-unittest,py37-unittest
+whitelist_externals =
+    poetry
```
Comment (abompard) [inline tox.ini:16]: Why not do `deps = poetry` instead?

File: fasjson/tests/fixtures/fasjson.cfg [hunk 19]
```diff
@@ -0,0 +1,3 @@
+KRB5CCNAME="/tmp/krb5cc-httpd"
```
Comment (abompard) [inline fasjson/tests/fixtures/fasjson.cfg:1]: This file used to have a `.env` extension because it's not a config file, it's a list of environment variables.

File: tox.ini (hunk 18 above)
Comment (StephenCoady) [reply tox.ini:16]: I tried that but it doesn't work. I had to install poetry via dnf first and then whitelist it. I can remove the whitelist from each individual tox env though

File: fasjson/tests/fixtures/fasjson.cfg (hunk 19 above)
Comment (StephenCoady) [reply fasjson/tests/fixtures/fasjson.cfg:1]: ok, reverting.

File: tox.ini (hunk 18 above)
Comment (abompard) [reply tox.ini]: Okay, yeah let's remove the additional whitelists

Comment (StephenCoady) [reply tox.ini]: yep, pushed in the last commit.
